pkg/maglev/maglev.go

```diff
@@ -132,3 +147,32 @@ func GetLookupTable(backends []string, m uint64) []int {
 
 	return entry
 }
+
+// derivePermutationSliceLen derives the permutations slice length depending on
+// the Maglev table size "m". The formula is (M / 100) * M * sizeof(uint64).
+// The heuristic is computes the slice size as follows:
+//
+//   251:    0.004806594848632812 MB
+//   509:    0.019766311645507812 MB
+//   1021:   0.07953193664550783 MB
+//   2039:   0.3171936798095703 MB
+//   4093:   1.2781256866455077 MB
+//   8191:   5.118750076293945 MB
+//   16381:  20.472500686645507 MB
+//   32749:  81.82502754211426 MB
+//   65521:  327.5300171661377 MB
+//   131071: 1310.700000076294 MB
+//
+// The heuristic does not apply to nodes with less than or equal to 8GB, as to
+// avoid memory pressure on memory-tight systems.
+//
+// Note, this function does not return the MB, but rather returns the number of
+// uint64 elements in the slice that equal to the total MB (length).
+func derivePermutationSliceLen(m uint64) uint64 {
+	threshold := uint64(8 * 1024 * 1024 * 1024) // 8GB
+	if vm, err := mem.VirtualMemory(); err != nil || vm == nil || vm.Total <= threshold {
+		return 0
+	}
+
+	return (m / uint64(100)) * m * uint64(unsafe.Sizeof(m))
```

Comment:
We don't need to multiply by `uint64(unsafe.Sizeof(m)`, as `make([]uint64, len)` will do that for us (it accepts number of elements which the function should return).

Reply:
Ah, you're right.